File: src/main/java/doom/despair/server/ClientHandler.java
package doom.despair.server;

import doom.despair.core.GameEvent;
import doom.despair.core.ServerGameEvent;
import doom.despair.core.events.ValidEvents;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class ClientHandler {
    private final BattleshipServer server;
    private final Socket socket;
    public ClientHandler(BattleshipServer server, Socket socket) {
        this.server = server;
        this.socket = socket;
    }
    public void handle() {
        try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(
                socket.getInputStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = bufferedReader.readLine()) != null) {
                try {
                    String[] s = line.split(";;");
                    int eventType = Integer.parseInt(s[0]);
                    Class<? extends ServerGameEvent> c = ValidEvents.SERVER_EVENTS.get(eventType);
                    if (c != null) {
                        ServerGameEvent event = GameEvent.deserialize(s[1], c);
                        server.queueEvent(event);
                    }
                } catch (NumberFormatException n) {
                    n.printStackTrace();
                }
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/doom/despair/server/ClientHandler.kt
package doom.despair.server

import doom.despair.core.Packet
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.Socket
import java.nio.charset.StandardCharsets

class ClientHandler(private val server: BattleshipServer, private val socket: Socket) {
    fun handle() {
        socket.use { clientSocket ->
            BufferedReader(
                InputStreamReader(
                    clientSocket.getInputStream(), StandardCharsets.UTF_8
                )
            ).use { reader ->
                BufferedWriter(
                    OutputStreamWriter(clientSocket.getOutputStream(), StandardCharsets.UTF_8)
                ).use { writer ->
                    processClient(reader, writer)
                }
            }
        }
    }

    private fun processClient(reader: BufferedReader, writer: BufferedWriter) {
        var registeredPlayerId: String? = null
        val sendPacket: (Packet) -> Unit = { packet ->
            synchronized(writer) {
                writer.write(packet.serialize())
                writer.newLine()
                writer.flush()
            }
        }
        try {
            var line: String?
            while ((reader.readLine().also { line = it }) != null) {
                val inputPacket = try {
                    Packet.deserialize(line!!)
                } catch (_: Exception) {
                    Packet(type = "ERROR", payload = """{"ok":false,"message":"Malformed packet"}""")
                }
                val response = server.processPacket(inputPacket, sendPacket)
                if (registeredPlayerId == null) {
                    registeredPlayerId = server.extractPlayerId(response)
                }
                sendPacket(response)
            }
        } catch (e: IOException) {
            if (server.isRunning) {
                throw RuntimeException(e)
            }
        } finally {
            registeredPlayerId?.let { server.unregisterSubscriber(it) }
        }
    }
}
